book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -167,6 +167,25 @@ chains, so that the map ordering is the ordering of best to worst chains.
 
 - XXX fill in details on exact types
 
+- **Chain**: `(im::OrdMap<block::Height, Arc<Block>>, HashSet<Nullifier>, HashSet<block::Hash>)`
+  - Ord impl is ordered by work
+  - push => add a block to the end of a chain, does contextual verification
+  checks, extracts info from block for extra data sets
+  - pop => remove the lowest block
+  - fork => create a new chain fork based on a given block(hash) within
+  another chain, calls push repeatedly
```

Comment:
Why does fork need to call push multiple times?
We only verify blocks one at a time, and we can't fork until the previous block has been verified.

It might help to make push and fork the same operation - but sometimes, the existing tip is replaced (push), and other times, a new tip is added (fork).

Reply:
In this design, the `Chain` type is supposed to represent a chain of blocks.  Each block is a state update, and the `Chain` type stores both the sequence of state updates (`BTreeMap<Height,Arc<Block>>`) as well as data that summarizes the state updates (nullifier sets, anchors, UTXOs, etc) over the whole sequence.

The `push` operation extends the chain with a new block, which can be done by updating the summary data (e.g., adding new nullifiers to the nullifier set).  The `pop` operation removes the root of the chain, which requires the opposite update to the summary data (e.g., removing nullifiers from the nullifier set).

The reason that `fork` is implemented as a sequence of push operations is that it's effectively rebuilding the summary data up to the fork point.  It's not possible to make push and fork the same operation without making the summary data per-block in some way, but this loses the advantage of having the summary data, which is that it can be checked per-chain rather than per block.

If the `pop` operation could pop from either end of the chain, it might be more efficient to implement `fork` by popping from the end of the chain down to the fork point, rather than rebuilding from the root, since forks are more likely to happen at the tip.